secboot/tpm.go

```diff
@@ -0,0 +1,209 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+// +build !nosecboot
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package secboot
+
+import (
+	"crypto/rand"
+	"fmt"
+
+	"github.com/chrisccoulson/go-tpm2"
+	sb "github.com/snapcore/secboot"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+const (
+	// Handles are in the block reserved for owner objects (0x01800000 - 0x01bfffff)
+	pinHandle = 0x01800000
+)
+
+type tpmSupport struct {
+	// Connection to the TPM device
+	tconn *sb.TPMConnection
+	// Lockout authorization
+	lockoutAuth []byte
+	// Paths to shim files
+	shimFiles []string
+	// Paths to bootloader files
+	bootloaderFiles []string
+	// Paths to kernel files
+	kernelFiles []string
+	// Kernel command lines
+	kernelCmdlines []string
+}
+
+func NewTPMSupport() (*tpmSupport, error) {
+	tconn, err := sb.ConnectToDefaultTPM()
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to default TPM: %v", err)
+	}
+
+	lockoutAuth := make([]byte, 16)
+	// crypto rand is protected against short reads
+	_, err = rand.Read(lockoutAuth)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create lockout authorization: %v", err)
+	}
+
+	t := &tpmSupport{tconn: tconn, lockoutAuth: lockoutAuth}
+
+	return t, nil
+}
+
+// StoreLockoutAuth saves the lockout authorization data in a file at the
+// path specified by filename.
+func (t *tpmSupport) StoreLockoutAuth(filename string) error {
```

Comment:
is this meant to be called before Provision or after usually?

Reply:
It's done before. We're trying to provision as late as possible to reduce the amount of things that could be wrong between provisioning and the end of the installation process. Currently only key sealing takes place after provisioning.

Reply:
ok, this should probably be renamed Save... in a follow up